### PR TITLE
feat: deno run --unstable-hmr

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds `--hmr` flag, that enabled Hot Module Replacement.

This flag works like `--watch` and accepts the same arguments. If
HMR is not possible the process will be restarted instead.

Currently HMR is only supported in `deno run` subcommand.

Upon HMR a `CustomEvent("hmr")` will be dispatched that contains
information which file was changed in its `details` property.

### CODE REVIEW

File: cli/tools/run/hot_reload/mod.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            let params = serde_json::from_value::<ScriptParsed>(notification.params)?;
+            if params.url.starts_with("file://") {
+              self.script_ids.insert(params.url, params.script_id);
+            }
+          }
+        }
+        changed_paths = self.path_change_receiver.recv() => {
+          for path in changed_paths?.iter().filter(|p| p.extension().map_or(false, |ext| ext == "js")) {
+            if let Some(path_str) = path.to_str() {
+              let module_url = "file://".to_owned() + path_str;
```
Comment (lucacasonato) [inline cli/tools/run/hot_reload/mod.rs:70]: `Url::from_file_path`

File: cli/tools/run/hot_reload/mod.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+              log::info!("{} Reloading changed module {}", colors::intense_blue("Hot-reload"), module_url);
+
+              if let Some(id) = self.script_ids.get(&module_url).map(String::clone) {
```
Comment (lucacasonato) [inline cli/tools/run/hot_reload/mod.rs:73]:
```suggestion
              if let Some(id) = self.script_ids.get(&module_url).cloned() {
```

**Comment by marvinhagemeister** [comment]
I tried to compile this branch, but ran into the following errors:

 ```
 error[E0624]: method `receive_from_v8_session` is private
   --> cli/tools/run/hot_reload/mod.rs:89:26
    |
89  |         _ = self.session.receive_from_v8_session() => {}
    |                          ^^^^^^^^^^^^^^^^^^^^^^^ private method
    |
   ::: /Users/marvinhagemeister/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deno_core-0.221.0/inspector.rs:804:3
    |
804 |   async fn receive_from_v8_session(&mut self) {
    |   ------------------------------------------- private method defined here

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> cli/tools/run/hot_reload/mod.rs:74:21
   |
74 |                 let src = tokio::fs::read_to_string(path).await?;
   |                     ^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
   = note: all local variables must have a statically known size
   = help: unsized locals are gated as an unstable feature

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> cli/tools/run/hot_reload/mod.rs:74:27
   |
74 |                 let src = tokio::fs::read_to_string(path).await?;
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
   = note: all local variables must have a statically known size
   = help: unsized locals are gated as an unstable feature

error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> cli/tools/run/hot_reload/mod.rs:74:64
   |
74 |                 let src = tokio::fs::read_to_string(path).await?;
   |                                                                ^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
note: required by a bound in `std::ops::ControlFlow::Break`
  --> /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/control_flow.rs:93:5

Some errors have detailed explanations: E0277, E0624.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `deno` (bin "deno") due to 4 previous errors
```

**Comment by bartlomieju** [comment]
@marvinhagemeister the error should go away if you tried to build after pulling this branch again.

**Comment by marvinhagemeister** [comment]
@bartlomieju Awesome, got it to work with the latest changes you pushed to the branch 👍  Played around with it and it's already amazing even in its early stages! It's soo much better than just having watch mode 💯 

Some notes:
- HMR seems to only work with `.js` files. We'd probably want to support the other extensions too `.ts`, `.tsx`, `.jsx`
- I really like that this only re-evaluates the current module and not the whole graph. Evaluating the whole graph breaks in scenarios where you have a `setInterval` running and print a value that you imported from another module. This just works in this PR 👍 
- HMR doesn't work when you change the value of a top level export like `export const foo = "asdf"`. This is totally fine and a fair tradeoff, but if hmr fails it should just print that it failed and do a full reload like it does in "normal" watch mode. Right now when it fails it just keeps running the previous code and won't accept any further updates anymore.
- Bonus: There might be a sneaky way to make top level ESM exports work by wrapping them in an iife like `export const foo = (() => "asdf")()` through transpilation. I've tried that locally and it accepts the update, but it doesn't re-evaluate the the parent module that imported this value. But then again even without this, the changes in the PR are already incredibly useful. This should in no way block merging this PR.
- Seems like this limitation exists for anything top-level. Changing something inside a function works.

So, in conclusion this is likely a fair tradeoff to make. From a product standpoint the only thing I'd add before shipping is:

- support `.ts`, `.jsx`, `.tsx`
- fall back to full reload similar to normal watch when hot update could not be applied. But still print the current error message so that the user is made aware of that.

**Comment by bartlomieju** [comment]
Thanks for trying this out!

> So, in conclusion this is likely a fair tradeoff to make. From a product standpoint the only thing I'd add before shipping is:
>
> - support .ts, .jsx, .tsx
> - fall back to full reload similar to normal watch when hot update could not be applied. But still print the current error message so that the user is made aware of that.

Very reasonable and I believe we just need a few hours of works to provide all this functionality.

In terms of DX - I really don't like `deno run --watch --hot-reload foo.js` - I think we should go with `deno run --hmr foo.js` - thoughts on this? How do we describe the difference between `--watch` and `--hmr`? Maybe `--watch` should just do HMR by default?

**Comment by marvinhagemeister** [comment]
> In terms of DX - I really don't like `deno run --watch --hot-reload foo.js` - I think we should go with `deno run --hmr foo.js` - thoughts on this? How do we describe the difference between `--watch` and `--hmr`? Maybe `--watch` should just do HMR by default?

I guess this depends on how confident we are in this. Long term it makes sense to combine the two. For now going with a separate `--watch` and `--hmr/--hot` flag seems like the safest option to go with. All bundlers I know of treat HMR as an opt-in thing separately from watch, so this would match already learned behavior.

My vote is for `deno run --hmr foo.js` or `deno run --hot foo.js`.

I think the difference can be described well:

- `--watch`: restarts the process whenever a file change is detected
- `--hmr`: reload the file that changed while keeping the process alive. If the HMR update cannot be applied it falls back to restarting the process

File: cli/args/flags.rs [hunk 3]
```diff
@@ -3810,6 +3821,7 @@ fn reload_arg_validate(urlstr: &str) -> Result<String, String> {
 fn watch_arg_parse(matches: &mut ArgMatches) -> Option<WatchFlags> {
   if matches.get_flag("watch") {
     Some(WatchFlags {
+      hot_reload: matches.get_flag("hot-reload"),
```
Comment (bartlomieju) [inline cli/args/flags.rs:3824]: Refactor this to not require `--watch` and `--hot` - only one of them should be required:
a) `--watch` - no HMR
b) `--hot` - `--watch` + `hot_reload: true`

File: cli/factory.rs [hunk 4]
```diff
@@ -63,21 +63,30 @@ use std::cell::RefCell;
...
 use std::path::PathBuf;
 use std::sync::Arc;
+use tokio::sync::broadcast::Receiver;
 
 pub struct CliFactoryBuilder {
-  maybe_sender: Option<tokio::sync::mpsc::UnboundedSender<Vec<PathBuf>>>,
+  maybe_watch_path_sender:
+    Option<tokio::sync::mpsc::UnboundedSender<Vec<PathBuf>>>,
+  maybe_changed_path_receiver:
+    Option<tokio::sync::broadcast::Receiver<Vec<PathBuf>>>,
```
Comment (bartlomieju) [inline cli/factory.rs:72]: I need to read through this part more, I'm not sure we need these two fields

File: cli/tools/bench/mod.rs [hunk 5]
```diff
@@ -421,7 +421,7 @@ pub async fn run_benchmarks_with_watch(
       let bench_flags = bench_flags.clone();
       Ok(async move {
         let factory = CliFactoryBuilder::new()
-          .with_watcher(sender.clone())
+          .with_watcher(sender.clone(), None)
```
Comment (bartlomieju) [inline cli/tools/bench/mod.rs:424]: I think I will refactor this so that `with_watcher()` accepts an options struct instead of two free args

File: cli/tools/run/hot_reload/json_types.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+pub struct SetScriptSourceReturnObject {
+  pub status: Status,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ScriptParsed {
+  pub script_id: String,
+  pub url: String,
+}
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/json_types.rs:20]: I believe some of these structs are coming from `Chrome Devtools Protocol` - if so I will move them to `cli/tools/repl/cdp.rs`

File: cli/tools/run/hot_reload/json_types.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+impl Status {
+  pub(crate) fn explain(&self) -> &'static str {
+    match self {
+      Status::Ok => "OK",
+      Status::CompileError => "compile error",
+      Status::BlockedByActiveGenerator => "blocked by active generator",
+      Status::BlockedByActiveFunction => "blocked by active function",
+      Status::BlockedByTopLevelEsModuleChange => {
+        "blocked by top-level ES module change"
+      }
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/json_types.rs:40]: These strings should be printed as a reason why we can't do HMR and require a full reload

File: cli/tools/run/hot_reload/mod.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      session,
+      path_change_receiver,
+      script_ids: HashMap::new(),
+    }
+  }
+
+  // FIXME(SyrupThinker): Inspector code duplication
+
+  pub async fn start(&mut self) -> Result<(), AnyError> {
+    self.enable_debugger().await?;
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:41]: This is duplicated in `cli/tools/coverage.rs` - I think I will abstract it away to a helper struct that wraps `LocalInspectorSession` and provides some methods to interact with the debugger

File: cli/tools/run/hot_reload/mod.rs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    Ok(())
+  }
+
+  pub async fn run(&mut self) -> Result<(), AnyError> {
+    let mut session_rx = self.session.take_notification_rx();
+    loop {
+      select! {
+        biased;
+        // TODO(SyrupThinker): Deferred retry with timeout
+        Some(notification) = session_rx.next() => {
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:58]: I will refactor this to not use `select!` and instead spawn a tokio task

File: cli/tools/run/hot_reload/mod.rs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+          let notification = serde_json::from_value::<RpcNotification>(notification)?;
+          if notification.method == "Debugger.scriptParsed" {
+            let params = serde_json::from_value::<ScriptParsed>(notification.params)?;
+            if params.url.starts_with("file://") {
+              self.script_ids.insert(params.url, params.script_id);
+            }
+          }
+        }
+        changed_paths = self.path_change_receiver.recv() => {
+          for path in changed_paths?.iter().filter(|p| p.extension().map_or(false, |ext| ext == "js")) {
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:68]: This bit needs to be changed, because it only reloads changes from JS files - we should support reloading `js`, `ts`, `tsx`, `jsx` and `json`.

File: cli/tools/run/hot_reload/mod.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/tools/run/hot_reload/mod.rs:73]: Also could use `let/else` to limit the indentation

File: cli/tools/run/mod.rs [hunk 11]
```diff
@@ -104,42 +106,80 @@ async fn run_with_watch(
...
-        worker.run_for_watcher().await?;
-
-        Ok(())
-      })
-    },
-  )
-  .await?;
-
-  Ok(0)
+  if watch_flags.hot_reload {
```
Comment (bartlomieju) [inline cli/tools/run/mod.rs:109]: This whole function might need to be refactored to support full restarts depending if HMR is possible or not

File: cli/worker.rs [hunk 12]
```diff
@@ -173,6 +189,12 @@ impl CliMainWorker {
         .with_event_loop(coverage_collector.stop_collecting().boxed_local())
         .await?;
     }
+    if let Some(hot_reload_manager) = maybe_hot_reload_manager.as_mut() {
+      self
+        .worker
+        .with_event_loop(hot_reload_manager.stop().boxed_local())
+        .await?;
+    }
```
Comment (bartlomieju) [inline cli/worker.rs:197]: If HMR manager stops here we should still allow for full reload when a file is changed.

File: cli/tools/run/hot_reload/json_types.rs (hunk 6 above)
Comment (marvinhagemeister) [reply cli/tools/run/hot_reload/json_types.rs:20]: yeah these look a lot like messages from CDP

**Comment by marvinhagemeister** [comment]
Coming off a talk with @bartlomieju and we chatted a bit about an API so that user code can be notified when an HMR update is applied. A use case for that is for Fresh where we'd like to notify the browser that the server applied an HMR update so that the browser can then initiate a new render with the updated content.

There is a bit of prior art on API designs for that in the vite ecosystem which went with a more modern spin on webpack's older HMR API. They went with `import.meta.hot.accept()` which is only available in HMR mode. Full docs here https://vitejs.dev/guide/api-hmr.html

For Deno though we both weren't sure if that's the right choice. VIte's API gets the module record passed as an argument which we don't have with the v8 API that we're basing this on. Instead a much more "web-like" alternative seems to instead go with a normal event dispatch instead. Something like:

```js
addEventListener("deno:hot", ev => {
  console.log(`[hmr] Update applied to ${ev.data.file}`)
  // prints: [hmr] Update applied to file://foo/bar/baz.tsx
});
```

A simple event listener would be quite a powerful API as we could hook it into lots of other things too like the test runner for example. For Fresh this would be perfect as well as this allows me to ping the browser and handle the update there.

File: cli/args/flags.rs
Comment (bartlomieju) [inline cli/args/flags.rs:1]: Update tests to use `--hmr` flag

File: cli/worker.rs (hunk 12 above)
Comment (bartlomieju) [reply cli/worker.rs:205]: This is tricky, due to timing. I introduced a "watcher mode" that depending on the variant either reloads whole process or wait for manual restart. The problem is that some programs (eg. `console.log("hello!")`) don't use TLA and thus finishing synchronously during evaluation. I think I might need to change the watcher modes to "hijack" watcher at some point instead of being fully manual.

**Comment by bartlomieju** [comment]
At this point, the bulk of this PR is touching file watcher code (related to `--watch` flag). I'm thinking of opening another PR that brings this changes first, so this PR becomes to review.

**Comment by marvinhagemeister** [comment]
Just tried out the latest changes in this PR and I can't get HMR to work with `.tsx` files. Steps to reproduce:

1. clone https://github.com/marvinhagemeister/deno-hmr-jsx
2. Run `deno run --hmr main.ts`
3. Change jsx text in `foo.tsx` -> change not detected

File: runtime/worker.rs [hunk 13]
```diff
@@ -523,6 +523,25 @@ impl MainWorker {
     }
   }
 
+  pub async fn with_event_loop_fallible<'a>(
```
Comment (bartlomieju) [inline runtime/worker.rs:526]: Try to revert this bit

File: cli/tools/run/mod.rs [hunk 14]
```diff
@@ -104,41 +107,83 @@ async fn run_with_watch(
...
```
Comment (bartlomieju) [inline cli/tools/run/mod.rs:110]: We should be able to unify these two arms now

File: cli/tools/run/hot_reload/mod.rs [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+            let result = hmr_manager.set_script_source(&id, source_code.as_str()).await?;
+
+            if matches!(result.status, Status::Ok) {
+              hmr_manager.dispatch_hmr_event(module_url.as_str()).await?;
+              break;
+            }
+
+            log::info!("{} Failed to reload module {}: {}.", colors::intense_blue("HMR"), module_url, colors::gray(result.status.explain()));
+            if !result.status.should_retry() {
+              log::info!("{} Restarting the process...", colors::intense_blue("HMR"));
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:222]: This should use `watcher_communicator` to print the message

File: cli/tools/run/hot_reload/mod.rs [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+          // TODO(bartlomieju): this loop should do 2 retries at most
+          loop {
+            let result = hmr_manager.set_script_source(&id, source_code.as_str()).await?;
+
+            if matches!(result.status, Status::Ok) {
+              hmr_manager.dispatch_hmr_event(module_url.as_str()).await?;
+              break;
+            }
+
+            log::info!("{} Failed to reload module {}: {}.", colors::intense_blue("HMR"), module_url, colors::gray(result.status.explain()));
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:220]: Ditto

File: cli/tools/run/hot_reload/mod.rs [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+          let Some(path_str) = path.to_str() else {
+            let _ = hmr_manager.watcher_communicator.force_restart();
+            continue;
+          };
+          let Ok(module_url) = Url::from_file_path(path_str) else {
+            let _ = hmr_manager.watcher_communicator.force_restart();
+            continue;
+          };
+
+          log::info!("{} Reloading changed module {}", colors::intense_blue("HMR"), module_url.as_str());
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:185]: Ditto

File: cli/worker.rs [hunk 18]
```diff
@@ -153,10 +165,28 @@ impl CliMainWorker {
...
+          hot_reload::run_hot_reload(hot_reload_manager).boxed_local();
+        let event_loop_future = self.worker.run_event_loop(false).boxed_local();
+
+        select! {
+          result = hmr_future => {
+            hot_reload_manager.watcher_communicator.change_restart_mode(WatcherRestartMode::Automatic);
+            result?;
+          },
+          result = event_loop_future => {
+            hot_reload_manager.watcher_communicator.change_restart_mode(WatcherRestartMode::Automatic);
```
Comment (bartlomieju) [inline cli/worker.rs:179]: Avoid this duplication here

File: cli/tools/run/hot_reload/mod.rs [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+          "contextId": Some(1),
+        })),
+      )
+      .await?;
+
+    Ok(())
+  }
+}
+
+// TODO(bartlomieju): Shouldn't use `tokio::select!` here, as futures are not cancel safe
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:133]: Remove this comment

File: cli/tools/run/hot_reload/mod.rs [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+          loop {
+            let result = hmr_manager.set_script_source(&id, source_code.as_str()).await?;
+
+            if matches!(result.status, Status::Ok) {
+              hmr_manager.dispatch_hmr_event(module_url.as_str()).await?;
+              break;
+            }
+
+            log::info!("{} Failed to reload module {}: {}.", colors::intense_blue("HMR"), module_url, colors::gray(result.status.explain()));
+            if !result.status.should_retry() {
```
Comment (bartlomieju) [inline cli/tools/run/hot_reload/mod.rs:221]: Add else here and sleep for 100ms, try twice and bail out.

**Comment by marvinhagemeister** [comment]
Tried the latest changes in various scenarios and it works amazingly well. I really love this feature, this is such a big improvement for DX 👍 

One minor nit I found is that the code frame visualization seems to lag 1 change behind when you caused an hmr change and throw an error in the next update:

Screenshot 2023-10-22 at 15 15 35

**Comment by marvinhagemeister** [comment]
Found a scenario where the process isn't restarted anymore after an error anymore:

```tsx
// foo.tsx
// Uncomment these lines and comment them again
// import foobar from "bar";
// console.log(foobar);

export function foo() {
  return `<h1>a</h1>`;
}

// main.ts
import { foo } from "./foo.tsx";

let i = 0;
setInterval(() => {
  console.log(i++, foo());
}, 1000);
```

Steps to reproduce:

1. Uncomment the invalid import in `foo.tsx` -> Error is correctly shown
2. Comment the invalid import again -> server is stuck

**Comment by bartlomieju** [comment]
Thanks for reporting that Marvin, both of these problems have been handled and are restarting properly 👍

File: cli/util/file_watcher.rs [hunk 21]
```diff
@@ -120,6 +148,10 @@ pub struct WatcherCommunicator {
 
   /// Send a message to force a restart.
   restart_tx: tokio::sync::mpsc::UnboundedSender<()>,
+
+  restart_mode: Arc<Mutex<WatcherRestartMode>>,
```
Comment (dsherret) [inline cli/util/file_watcher.rs:152]: Would it be better if we put an Arc around the entire WatcherCommunicator? (or actually looks like we might be able to do Rc?)

File: cli/util/file_watcher.rs [hunk 22]
```diff
@@ -206,19 +260,41 @@ where
...
+    restart_mode: restart_mode.clone(),
+    banner: colors::intense_blue(banner).to_string(),
   };
-  info!("{} {} started.", colors::intense_blue("Watcher"), job_name,);
+  info!("{} {} started.", colors::intense_blue(banner), job_name);
+
+  let changed_paths = Arc::new(Mutex::new(None));
+  let changed_paths_ = changed_paths.clone();
+
+  tokio::task::spawn(async move {
```
Comment (dsherret) [inline cli/util/file_watcher.rs:282]: Use spawn from deno_core? Then use Rc::RefCell above? Actually, does `restart_mode` need to be in a Mutex or can it be in a RefCell?

File: cli/worker.rs [hunk 23]
```diff
@@ -153,10 +165,33 @@ impl CliMainWorker {
...
+            result = hmr_result;
+          },
+          event_loop_result = event_loop_future => {
+            result = event_loop_result;
+          }
+        }
+        if let Err(e) = result {
+          hot_reload_manager
+            .watcher_communicator
+            .change_restart_mode(WatcherRestartMode::Automatic);
```
Comment (dsherret) [inline cli/worker.rs:185]: Line to refactor somehow to not access a dependency of a dependency.

Example:

```
 let communicator =
      self.shared.maybe_file_watcher_communicator.as_ref().unwrap();
 communicator.change_restart_mode(WatcherRestartMode::Automatic);
```

Or perhaps `maybe_hot_reload_manager` could contain the communicator as well.

File: cli/worker.rs [hunk 24]
```diff
@@ -287,6 +328,32 @@ impl CliMainWorker {
...
   }
 
+  pub async fn maybe_setup_hot_reload_manager(
+    &mut self,
+  ) -> Result<Option<HotReloadManager>, AnyError> {
+    if !self.shared.options.hot_reload {
+      return Ok(None);
+    }
+
+    let interface =
```
Comment (dsherret) [inline cli/worker.rs:338]:
```suggestion
    let communicator =
```

File: cli/tools/run/hot_reload/mod.rs [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+use tokio::select;
+
+mod json_types;
+
+use json_types::RpcNotification;
+use json_types::ScriptParsed;
+use json_types::SetScriptSourceReturnObject;
+use json_types::Status;
+
+pub struct HotReloadManager {
```
Comment (dsherret) [inline cli/tools/run/hot_reload/mod.rs:26]: We don't really have anything named `Manager` in the codebase (and I always think of [this blog post](https://blog.codinghorror.com/i-shall-call-it-somethingmanager/) when I see it). Would be nice to think of a better name.

**Review by dsherret — COMMENTED** [review]
This will be a nice feature.

File: cli/tools/run/hot_reload/mod.rs [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+        })),
+      )
+      .await?;
+
+    Ok(())
+  }
+}
+
+pub async fn run_hot_reload(
+  hmr_manager: &mut HotReloadManager,
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/tools/run/hot_reload/mod.rs:133]: It was that originally, but I prefer to have an explicitly function over a method, because it needs to be run in `select!` macro and there's too much coupling then

File: cli/util/file_watcher.rs (hunk 21 above)
Comment (bartlomieju) [reply cli/util/file_watcher.rs:152]: I'm not sure it would help much as `changed_paths_rx` requires an explicit function call to create a "clone" of it. `restart_mode` is the only field that requires internal mutability.

File: cli/util/file_watcher.rs (hunk 22 above)
Comment (bartlomieju) [reply cli/util/file_watcher.rs:282]: If a blocking loop is run in JS code then the file changes would never be discovered (as the blocking loop would never yield to the Tokio runtime). I can add a comment here

File: cli/worker.rs (hunk 23 above)
Comment (bartlomieju) [reply cli/worker.rs:185]: But `maybe_hot_reload_manager` already contains is as `hot_reload_manager.watcher_communicator`. I can change it to use `self.shared.maybe_file_watcher_communicator` though.

File: cli/tools/run/hot_reload/mod.rs (hunk 25 above)
Comment (bartlomieju) [reply cli/tools/run/hot_reload/mod.rs:26]: Sure; I'll change the name.

File: cli/emit.rs [hunk 27]
```diff
@@ -101,6 +101,28 @@ impl Emitter {
...
+  pub async fn load_and_emit_for_hmr(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<String, AnyError> {
+    let media_type = MediaType::from_specifier(specifier);
+    let source_code = tokio::fs::read_to_string(
+      ModuleSpecifier::to_file_path(specifier).unwrap(),
+    )
+    .await?;
+    let source_arc: Arc<str> = Arc::from(source_code.as_str());
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/emit.rs:114]: I don't think it clones under the hood, `Arc::from(source_code.into())` doesn't work here.

[the remark this answers is not in the document]
Comment (dsherret) [reply cli/emit.rs:114]: I mean, try doing this:

```
let source_arc: Arc<str> = source_code.into()
```

I think maybe `.as_str()` does clone because I can use `source_code` after that line?

File: cli/tools/run/hot_reload/mod.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/tools/run/hot_reload/mod.rs]: Don't they both have the same amount of coupling though and aren't they essentially equivalent other than needing to do an additional import for the first one? What do you mean by the select! macro?

File: cli/util/file_watcher.rs [hunk 28]
```diff
@@ -206,19 +262,41 @@ where
...
     paths_to_watch_tx: paths_to_watch_tx.clone(),
     changed_paths_rx: changed_paths_rx.resubscribe(),
     restart_tx: restart_tx.clone(),
+    restart_mode: restart_mode.clone(),
+    banner: colors::intense_blue(banner).to_string(),
   };
-  info!("{} {} started.", colors::intense_blue("Watcher"), job_name,);
+  info!("{} {} started.", colors::intense_blue(banner), job_name);
+
+  let changed_paths = Arc::new(Mutex::new(None));
```
Comment (dsherret) [inline cli/util/file_watcher.rs:281]: This can be an Rc::new(RefCell::new(None))

File: cli/util/file_watcher.rs (hunk 21 above)
Comment (dsherret) [reply cli/util/file_watcher.rs:154]: I just tried it out and it works to remove the `Clone` implementation then just pass around an `Arc<WatcherCommunicator>` everywhere. It's much more clear that it's the same instance being passed around when it's in an `Arc`.

**Review by dsherret — APPROVED** [review]
LGTM!
